### PR TITLE
Passthrough autopilot - added ports array case and updated unit tests

### PR DESCRIPTION
**Pull request by vicentefb**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation
/kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:
Handles case where we receive an array of mixed ports (Dynamic and Passthrough) for Autopilot clusters. 

**Which issue(s) this PR fixes**:

Towards #3721

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/apis/agones/v1/gameserver.go [hunk 1]
```diff
@@ -156,6 +156,8 @@ const (
 	// NodePodIP identifies an IP address from a pod.
 	NodePodIP corev1.NodeAddressType = "PodIP"
 
+	PassthroughPortAssignmentAnnotation = "autopilot.gke.io/container-passthrough-port-assignment"
```
Comment (zmerlynn) [inline pkg/apis/agones/v1/gameserver.go:160]: Use `agones.dev` rather than `autopilot.gke.io`. (In general, annotations are "owned" by the domain name - so it would be surprising to find the Agones project emitting a `gke.io` annotation.)

File: pkg/apis/agones/v1/gameserver_test.go [hunk 2]
```diff
@@ -1497,6 +1498,48 @@ func TestGameServerDisableServiceAccount(t *testing.T) {
...
+	require.NoError(t, runtime.ParseFeatures(string(runtime.FeatureAutopilotPassthroughPort)+"=true"))
+	gs := &GameServer{ObjectMeta: metav1.ObjectMeta{Name: "gameserver", UID: "1234"}, Spec: GameServerSpec{
+		Ports: []GameServerPort{
+			{Name: "defaultPassthrough", PortPolicy: Passthrough},
+			{Name: "defaultDynamic", PortPolicy: Dynamic, ContainerPort: 7654},
+			{Name: "defaultDynamicTwo", PortPolicy: Dynamic, ContainerPort: 7659},
+			{Name: "defaultPassthroughTwo", PortPolicy: Passthrough},
+		},
+		Template: corev1.PodTemplateSpec{
+			Spec: corev1.PodSpec{
```
Comment (zmerlynn) [inline pkg/apis/agones/v1/gameserver_test.go:1514]: Please test the multi-container case. The `Ports` array can take a `container` name as well, so `Ports` could be a mix of different container ports.

File: pkg/apis/agones/v1/gameserver.go [hunk 3]
```diff
@@ -745,9 +747,10 @@ func (gs *GameServer) Pod(apiHooks APIHooks, sidecars ...corev1.Container) (*cor
 	}
 
 	gs.podObjectMeta(pod)
-	for _, p := range gs.Spec.Ports {
-		var hostPort int32
 
+	passthroughContainerPortMap := make(map[string][]string)
```
Comment (zmerlynn) [inline pkg/apis/agones/v1/gameserver.go:752]: why string and not int?

File: pkg/gameservers/controller.go [hunk 4]
```diff
@@ -334,8 +334,27 @@ func (ext *Extensions) creationMutationHandlerPod(review admissionv1.AdmissionRe
 
 	ext.baseLogger.WithField("pod.Name", pod.Name).Debug("creationMutationHandlerPod")
 
-	// TODO: We need to deal with case of multiple and mixed type ports before enabling the feature gate.
-	pod.Spec.Containers[1].Ports[0].ContainerPort = pod.Spec.Containers[1].Ports[0].HostPort
+	annotation, ok := pod.ObjectMeta.Annotations[agonesv1.PassthroughPortAssignmentAnnotation]
+	if !ok {
+		return review, nil
```
Comment (zmerlynn) [inline pkg/gameservers/controller.go:339]: log this case? We're not expecting to see it, are we, since we have an object selector that limits it to the specific label for passthrough? So we know it has at least one port.

File: pkg/util/runtime/features.go [hunk 5]
```diff
@@ -140,7 +140,7 @@ var (
 		FeaturePortPolicyNone:                   false,
 
 		// Dev features
-		FeatureAutopilotPassthroughPort: false,
```
Comment (zmerlynn) [inline pkg/util/runtime/features.go]: Did you mean to include this? This would be the equivalent of promoting it to Beta.

I think after this change goes in you can move it up to Alpha, though (follow the instructions above to do that).

**Review by zmerlynn — COMMENTED** [review]
Looking good! A few comments.

File: pkg/apis/agones/v1/gameserver.go (hunk 1 above)
Comment (vicentefb) [reply pkg/apis/agones/v1/gameserver.go:160]: Okok sounds good!

File: pkg/apis/agones/v1/gameserver.go (hunk 3 above)
Comment (vicentefb) [reply pkg/apis/agones/v1/gameserver.go:752]: True, it can be an int and that will save some conversions.

File: pkg/gameservers/controller.go (hunk 4 above)
Comment (vicentefb) [reply pkg/gameservers/controller.go:339]: You're right, i'll add a debug log, thanks!

File: pkg/apis/agones/v1/gameserver_test.go (hunk 2 above)
Comment (vicentefb) [reply pkg/apis/agones/v1/gameserver_test.go:1521]: Added!

File: pkg/gameservers/controller.go [hunk 6]
```diff
@@ -334,8 +334,24 @@ func (ext *Extensions) creationMutationHandlerPod(review admissionv1.AdmissionRe
...
+	passthroughPortAssignmentMap := make(map[string][]int)
+	if err := json.Unmarshal([]byte(annotation), &passthroughPortAssignmentMap); err != nil {
+		return review, errors.Wrapf(err, "could not unmarshal annotation %q (value %q)", passthroughPortAssignmentMap, annotation)
+	}
+
+	for _, p := range pod.Spec.Containers {
+		if containerPassthrough, ok := passthroughPortAssignmentMap[p.Name]; ok {
+			for i := range containerPassthrough {
+				i := containerPassthrough[i]
+				p.Ports[i].ContainerPort = p.Ports[i].HostPort
```
Comment (zmerlynn) [inline pkg/gameservers/controller.go:351]: So, this is subtle but I think your logic here is off. In particular, in gameserver.go you're recording the index within the Ports array, but here you're indexing into container.Ports by the _absolute_ index, vs the index within that container. You can see this in your new unit text where the annotation records `containerFour: [5]`, but that's index 5 into the GameServer Ports array, not index 5 into the container Ports array.

To fix this, I'd recommend two things:
* In gamserver.go, instead of adding an index variable to the range over the GameServer ports, record the index before the line `c.Ports = append(c.Ports, cp)`, I.e. by `portIdx := len(c.Ports)`.
* Test the multi-container case here as well - in fact if you mirror the container/port structure in the test above you just added, it'll make for a good symmetric test.

File: pkg/gameservers/controller.go [hunk 7]
```diff
@@ -334,8 +334,24 @@ func (ext *Extensions) creationMutationHandlerPod(review admissionv1.AdmissionRe
...
+	if !ok {
+		ext.baseLogger.WithField("pod.Name", pod.Name).Debug("creationMutationHandlerPod, " + agonesv1.PassthroughPortAssignmentAnnotation + ": " + pod.ObjectMeta.Annotations[agonesv1.PassthroughPortAssignmentAnnotation])
+	}
+
+	passthroughPortAssignmentMap := make(map[string][]int)
+	if err := json.Unmarshal([]byte(annotation), &passthroughPortAssignmentMap); err != nil {
+		return review, errors.Wrapf(err, "could not unmarshal annotation %q (value %q)", passthroughPortAssignmentMap, annotation)
+	}
+
+	for _, p := range pod.Spec.Containers {
```
Comment (zmerlynn) [inline pkg/gameservers/controller.go:347]: total nit but `p` is a weird name for a container. `c`? It confused me when I saw `p.Ports[i]` because I thought it was somehow an array on the pod.

File: pkg/gameservers/controller.go [hunk 8]
```diff
@@ -334,8 +334,24 @@ func (ext *Extensions) creationMutationHandlerPod(review admissionv1.AdmissionRe
...
+		if containerPassthrough, ok := passthroughPortAssignmentMap[p.Name]; ok {
+			for i := range containerPassthrough {
+				i := containerPassthrough[i]
```
Comment (zmerlynn) [inline pkg/gameservers/controller.go:350]: Two things:
1) line 349/350 are a weird pairing because you're ranging over the indices, then immediately shadowing with the value. This can be shortened to just: `for _, i := range containerPassthrough`

That said, though, I would shy away from `i` here and probably make it slightly more descriptive, like `portIdx`.

2) Fun fact, `passthroughPortAssignmentMap[p.Name]` will default to `nil`, and a `nil` slice is an empty slice, so you can skip 348.

Basically I think line 348-350 can be replaced with:
```
for _, portIdx := range passthroughPortAssignmentMap[p.Name] {
}
```

**Review by zmerlynn — COMMENTED** [review]
Closer!

File: pkg/gameservers/controller.go (hunk 8 above)
Comment (vicentefb) [reply pkg/gameservers/controller.go]: Okok sg, I changed the for loop variable name and replaed lines 348-350.

File: pkg/gameservers/controller.go (hunk 6 above)
Comment (vicentefb) [reply pkg/gameservers/controller.go]: Ohhhh interesting, good catch! I made the changes and now i'm using the container port index instead and updated unit test to reflect this mirror in the multi container test, thanks!!

File: pkg/gameservers/controller.go [hunk 9]
```diff
@@ -334,8 +334,21 @@ func (ext *Extensions) creationMutationHandlerPod(review admissionv1.AdmissionRe
 
 	ext.baseLogger.WithField("pod.Name", pod.Name).Debug("creationMutationHandlerPod")
 
-	// TODO: We need to deal with case of multiple and mixed type ports before enabling the feature gate.
-	pod.Spec.Containers[1].Ports[0].ContainerPort = pod.Spec.Containers[1].Ports[0].HostPort
+	annotation, ok := pod.ObjectMeta.Annotations[agonesv1.PassthroughPortAssignmentAnnotation]
+	if !ok {
+		ext.baseLogger.WithField("pod.Name", pod.Name).Debug("creationMutationHandlerPod, " + agonesv1.PassthroughPortAssignmentAnnotation + ": " + pod.ObjectMeta.Annotations[agonesv1.PassthroughPortAssignmentAnnotation])
```
Comment (zmerlynn) [inline pkg/gameservers/controller.go:339]: I don't think you need to log the value here because you know it's empty. Just say that the annotation is empty and it's unexpected. Feel free to bump it to an `Info`, too, since it's unexpected (so potentially more relevant than debug), but not really that important as it's just noise.

Separately I think you still want to return here after logging, no? I presume the unmarshal will fail o/w.

File: pkg/gameservers/controller_test.go [hunk 10]
```diff
@@ -2275,15 +2277,32 @@ func newSingleContainerSpec() agonesv1.GameServerSpec {
...
+	containerToPassthroughMapJSON, err := json.Marshal(passthroughContainerPortMap)
+	if err != nil {
+		return corev1.Pod{}
+	}
+	return corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{agonesv1.PassthroughPortAssignmentAnnotation: string(containerToPassthroughMapJSON)},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
```
Comment (zmerlynn) [inline pkg/gameservers/controller_test.go:2295]: I thought you were going to change this test to also be multi-container?

File: pkg/gameservers/controller_test.go [hunk 11]
```diff
@@ -2275,15 +2277,32 @@ func newSingleContainerSpec() agonesv1.GameServerSpec {
...
-				Image: "container2/image",
-				Ports: []corev1.ContainerPort{{HostPort: 7777, ContainerPort: 555}},
-				Env:   []corev1.EnvVar{{Name: passthroughPortEnvVar, Value: "TRUE"}}}},
+// Assume ports 0 and 2 are Passthrough ports for "example-server" container
+// The annotation would look like autopilot.gke.io/passthrough-port-assignment: '{"example-server":["0","2"]}'
+func newPassthroughPortSingleContainerSpec() corev1.Pod {
+	passthroughContainerPortMap := make(map[string][]int)
+	passthroughContainerPortMap["example-server"] = append(passthroughContainerPortMap["example-server"], 0, 2)
+	var containerToPassthroughMapJSON []byte
+	containerToPassthroughMapJSON, err := json.Marshal(passthroughContainerPortMap)
```
Comment (zmerlynn) [inline pkg/gameservers/controller_test.go:2286]: As much as it seems like it's more readable to test it using a map, you expose yourself to the possibility of a "symmetric error" since you're doing this same operation to construct the JSON in both tests. It would be better just to use a string here, and it also documents the format.

File: pkg/gameservers/controller_test.go [hunk 12]
```diff
@@ -2275,15 +2277,32 @@ func newSingleContainerSpec() agonesv1.GameServerSpec {
...
-		Containers: []corev1.Container{
-			{Name: "agones-gameserver-sidecar",
-				Image: "container/image",
-				Env:   []corev1.EnvVar{{Name: passthroughPortEnvVar, Value: "TRUE"}}},
-			{Name: "simple-game-server",
-				Image: "container2/image",
-				Ports: []corev1.ContainerPort{{HostPort: 7777, ContainerPort: 555}},
-				Env:   []corev1.EnvVar{{Name: passthroughPortEnvVar, Value: "TRUE"}}}},
+// Assume ports 0 and 2 are Passthrough ports for "example-server" container
+// The annotation would look like autopilot.gke.io/passthrough-port-assignment: '{"example-server":["0","2"]}'
```
Comment (zmerlynn) [inline pkg/gameservers/controller_test.go:2281]: I think this comment is out of date (it's `{"example-server":[0,2]}`, right?)

**Review by zmerlynn — COMMENTED** [review]
Almost!

File: pkg/apis/agones/v1/gameserver.go [hunk 13]
```diff
@@ -745,9 +748,11 @@ func (gs *GameServer) Pod(apiHooks APIHooks, sidecars ...corev1.Container) (*cor
 	}
 
 	gs.podObjectMeta(pod)
+
+	passthroughContainerPortMap := make(map[string][]int)
+	portIdx := 0
```
[the remark this answers is not in the document]
Comment (vicentefb) [reply pkg/apis/agones/v1/gameserver.go:753]: I need `passthroughContainerPortMap` to be outside the for loop because i use it in the following if statement from line 779.

[the remark this answers is not in the document]
Comment (zmerlynn) [reply pkg/apis/agones/v1/gameserver.go:753]: but `portIdx` only applies inside the loop, right?

File: pkg/gameservers/controller_test.go (hunk 12 above)
Comment (vicentefb) [reply pkg/gameservers/controller_test.go:2281]: I think it could still be valid case unless I'm missing something (?) In this method i'm building the PodSpec and the `example-server` container has 3 ports so technically container ports 0 and 2 are still valid. I'm aware that the `HostPort` and `ContainerPort` are not the same value in this piece of code but that's intentional since at this point the webhook hasn't been triggered. Nevertheless, I included another container in this PodSpec and updated the comment.

File: pkg/apis/agones/v1/gameserver.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (vicentefb) [reply pkg/apis/agones/v1/gameserver.go]: Yeah, I did move `portIdx` to happen inside :)

**Review by zmerlynn — APPROVED** [review]

**Comment by zmerlynn** [comment]
Great job!
